jbmc/src/java_bytecode/remove_java_new.cpp

```diff
@@ -153,22 +153,50 @@ goto_programt::targett remove_java_newt::lower_java_new_array(
 
   const struct_typet &struct_type = to_struct_type(ns.follow(object_type));
 
-  // Ideally we would have a check for `is_valid_java_array(struct_type)` but
-  // `is_valid_java_array is part of the java_bytecode module and we cannot
-  // introduce such dependencies. We do this simple check instead:
-  PRECONDITION(struct_type.components().size() == 3);
+  PRECONDITION(is_valid_java_array(struct_type));
 
   // Init base class:
   dereference_exprt deref(lhs, object_type);
   auto zero_object = zero_initializer(object_type, location, ns);
   CHECK_RETURN(zero_object.has_value());
-  set_class_identifier(
-    to_struct_expr(*zero_object), ns, to_struct_tag_type(object_type));
+  set_class_identifier(to_struct_expr(*zero_object), ns, object_type);
   dest.insert_before(
     next,
     goto_programt::make_assignment(
       code_assignt(deref, *zero_object), location));
 
+  // If it's a reference array we need to set the dimension and element type
+  // fields (if the given type is void then the caller must set these up itself;
+  // this happens in array[reference].clone(), where the type info must be
+  // copied over from the input array)
+  const auto underlying_type_and_dimension =
+    java_reference_array_dimension_and_element_type(object_type);
+
+  if(
+    underlying_type_and_dimension.second >= 1 &&
```

Comment:
:pick: might be even clearer to introduce a tempoary

```
const bool is_reference_array  = underlying_type_and_dimension.second >= 1
```

Reply:
That's actually not accurate, as an array-of-`void*` passes that test, as per the comment above